plugins/flatpak/gs-flatpak.c

```diff
@@ -2647,19 +2742,18 @@ gs_flatpak_get_list_for_install (GsFlatpak *self, GsApp *app,
 	if (!gs_flatpak_refine_runtime_for_install (self, app, cancellable, error))
 		return FALSE;
 	runtime = gs_app_get_update_runtime (app);
-	if (runtime != NULL) {
-		g_autofree gchar *ref_display = NULL;
-		ref_display = gs_flatpak_app_get_ref_display (runtime);
-		if (g_hash_table_contains (hash_installed, ref_display)) {
-			g_debug ("%s is already installed, so skipping",
-				 gs_app_get_id (runtime));
-		} else {
-			g_debug ("%s/%s is not already installed, so installing",
-				 gs_flatpak_app_get_ref_name (runtime),
-				 gs_flatpak_app_get_ref_branch (runtime));
-			gs_app_list_add (list, runtime);
-		}
+	if (runtime != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (runtime, list, hash_installed);
+
+	/* add services flatpak */
+	services = gs_flatpak_get_services_app_for_runtime (self, runtime, cancellable);
+	if (g_cancellable_is_cancelled (cancellable)) {
```

Comment:
You can use `g_cancellable_set_error_if_cancelled` which sets the error and returns `TRUE` if it's been cancelled. So all you'd need to do is return FALSE in that case.